src/net/ossl.cc

```diff
@@ -859,13 +859,119 @@ class session : public enable_shared_from_this<session>, public session_impl {
         using result_t = std::vector<subject_alt_name>;
-        return make_exception_future<result_t>(std::runtime_error("unimplemented"));
+
+        if (_error) {
+            return make_exception_future<result_t>(_error);
+        }
+        if (_shutdown) {
+            return make_exception_future<result_t>(std::system_error(ENOTCONN, std::system_category()));
+        }
+        if (!connected()) {
+            return handshake().then([this, types = std::move(types)]() mutable {
+               return get_alt_name_information(std::move(types));
+            });
+        }
+
+        auto peer_cert = get_peer_certificate();
+        if (!peer_cert) {
+            return make_ready_future<result_t>();
+        }
+        return make_ready_future<result_t>(do_get_alt_name_information(peer_cert, types));
     }
 
 
 private:
+    std::vector<subject_alt_name> do_get_alt_name_information(std::shared_ptr<X509> &peer_cert,
+                                                              const std::unordered_set<subject_alt_name_type> &types) const {
+        int ext_idx = X509_get_ext_by_NID(peer_cert.get(), NID_subject_alt_name, -1);
+        if (ext_idx >= 0) {
+            X509_EXTENSION *ext = X509_get_ext(peer_cert.get(), ext_idx);
+            if (ext) {
+                GENERAL_NAMES *names = (GENERAL_NAMES*)X509V3_EXT_d2i(ext);
+                if (names) {
+                    int num_names = sk_GENERAL_NAME_num(names);
+                    std::vector<subject_alt_name> alt_names;
+                    alt_names.reserve(num_names);
+
+                    for (auto i = 0; i < num_names; i++) {
+                        GENERAL_NAME *name = sk_GENERAL_NAME_value(names, i);
+                        if (auto known_t = field_to_san_type(name)) {
+                            if (types.empty() || types.count(known_t->type)) {
+                                alt_names.push_back(std::move(*known_t));
+                            }
+                        }
+                    }
+                    return alt_names;
+                }
+            }
+        }
+        return {};
+    }
+
+    std::optional<subject_alt_name> field_to_san_type(GENERAL_NAME* name) const {
+        subject_alt_name san;
+        switch(name->type) {
+            case GEN_IPADD:
+            {
+                union {
+                    char c;
+                    ::in_addr in;
+                    ::in6_addr in6;
+                } tmp;
+                san.type = subject_alt_name_type::ipaddress;
+                const auto size = ASN1_STRING_length(name->d.iPAddress);
+                memcpy(&tmp.c, ASN1_STRING_get0_data(name->d.iPAddress), size);
+                if (size == sizeof(::in_addr)) {
+                    san.value = net::inet_address(tmp.in);
+                } else if (size == sizeof(::in6_addr)) {
+                    san.value = net::inet_address(tmp.in6);
```

Comment:
Pretty sure this fails strict-aliasing.

If seastar requires C++20 (not sure if that's happened yet), then  you should use a `std::array<char>` and `std::bit_cast`.